moveit_ros/move_group/src/default_capabilities/execute_trajectory_action_capability.cpp

```diff
@@ -47,90 +47,183 @@ MoveGroupExecuteTrajectoryAction::MoveGroupExecuteTrajectoryAction() : MoveGroup
   }
+}
 
-  setExecuteTrajectoryState(IDLE);
+void MoveGroupExecuteTrajectoryAction::cancelCallback(ExecuteTrajectoryActionServer::GoalHandle goal_handle)
+{
+  ROS_DEBUG_STREAM_NAMED(getName(),
+                         "Cancel goal requested (ExecuteTrajectoryActionServer): " << goal_handle.getGoalID());
+  context_->trajectory_execution_manager_->stopExecution(true);  // TODO: fix
+  const std::string response = "This goal was canceled by the user";
+  cancelGoal(goal_handle, response);
 }
 
-void MoveGroupExecuteTrajectoryAction::executePath(const moveit_msgs::ExecuteTrajectoryGoalConstPtr& goal,
-                                                   moveit_msgs::ExecuteTrajectoryResult& action_res)
+void MoveGroupExecuteTrajectoryAction::executePath(ExecuteTrajectoryActionServer::GoalHandle goal_handle)
 {
-  ROS_INFO_NAMED(getName(), "Execution request received");
+  if (goal_handle.getGoalStatus().status != actionlib_msgs::GoalStatus::PENDING)
+  {
+    ROS_INFO_NAMED(getName(), "Preempt requested before the goal is accepted");
+    return;
+  }
+
+  goal_handle.setAccepted("This goal has been accepted by the action server");
+
+  if (!context_->trajectory_execution_manager_->getEnableSimultaneousExecution())
+    context_->trajectory_execution_manager_->clear();
+
+  auto executionCallback = [this, goal_handle](const moveit_controller_manager::ExecutionStatus& status) {
+    sendGoalResponse(goal_handle, status);
+  };
 
-  context_->trajectory_execution_manager_->clear();
-  if (context_->trajectory_execution_manager_->push(goal->trajectory))
+  if (context_->trajectory_execution_manager_->push(goal_handle.getGoal()->trajectory, "", executionCallback))
   {
-    setExecuteTrajectoryState(MONITOR);
-    context_->trajectory_execution_manager_->execute();
-    moveit_controller_manager::ExecutionStatus status = context_->trajectory_execution_manager_->waitForExecution();
-    if (status == moveit_controller_manager::ExecutionStatus::SUCCEEDED)
-    {
-      action_res.error_code.val = moveit_msgs::MoveItErrorCodes::SUCCESS;
-    }
-    else if (status == moveit_controller_manager::ExecutionStatus::PREEMPTED)
-    {
-      action_res.error_code.val = moveit_msgs::MoveItErrorCodes::PREEMPTED;
-    }
-    else if (status == moveit_controller_manager::ExecutionStatus::TIMED_OUT)
-    {
-      action_res.error_code.val = moveit_msgs::MoveItErrorCodes::TIMED_OUT;
-    }
-    else
-    {
-      action_res.error_code.val = moveit_msgs::MoveItErrorCodes::CONTROL_FAILED;
-    }
-    ROS_INFO_STREAM_NAMED(getName(), "Execution completed: " << status.asString());
+    setExecuteTrajectoryState(MONITOR, goal_handle);
+    if (!context_->trajectory_execution_manager_->getEnableSimultaneousExecution())
+      context_->trajectory_execution_manager_->execute(executionCallback, true);
   }
   else
   {
-    action_res.error_code.val = moveit_msgs::MoveItErrorCodes::CONTROL_FAILED;
+    ROS_ERROR_NAMED(getName(), "Failed to pushed trajectory");
```

Comment:
```suggestion
    ROS_ERROR_NAMED(getName(), "Failed to push trajectory");
```
when and why does it fail? This error message is not super helpful

Reply:
Okay, I removed it. There is no way to know why it failed here so the log message should come from the TrajectoryExecutionManager.

Reply:
Right. I removed the log message here as there is no way to know why it failed. The TrajectoryExecutionManager should log a more informative message.